### ISSUE TITLE
should we further mitigate threats of identifiers in local storage because some mechanisms won't clear all local storage?

### ISSUE THREAD

**Issue by npdoty** [w3ctag/security-questionnaire#80]
In this questionnaire and in the mitigating browser fingerprinting guidance, we note that local storage mechanisms should be identified for simultaneous clearing to avoid evercookies or other persistent state.

@pes10k noted a concern about unique identifiers, even if cleared with other local storage, because some user mitigation mechanisms (like privacy-focused browser extensions or automatic expiration of cookies) might not clear them. This issue is to track that question and see if we should update guidance that simultaneous clearing of local storage isn't enough. Pete, if you can add any more details on the threat/use cases, that would be helpful.

**Comment by npdoty** [comment]
I note that we do have a positive example in the questionnaire about the Gamepad API, which uses an incrementing integer for its index of devices, rather than a unique identifier: https://w3ctag.github.io/security-questionnaire/#temporary-id

And the section on persistent state also discusses the mitigation of clearing all local storage simultaneously:
https://w3ctag.github.io/security-questionnaire/#persistent-origin-specific-state

**Comment by annevk** [comment]
At least specification-wise this is being addressed by rewriting all storage APIs to be on top of storage buckets, as defined by https://storage.spec.whatwg.org/.

**Comment by pes10k** [comment]
@annevk im not sure I understand how the buckets definition addresses the concern.  A privacy intervention like "cap all JS cookies to N days lifetime" wouldn't seem to trigger that definition, since it seems like the trigger there is a browser / affect that wants to clear the entire bucket.

The broader concern is that there are privacy mitigations being pursued by vendors that have the goal of reducing how much "stuff" gets stored (e.g. life timing keys / cookies / fields, "garbage collecting" rarely used items, etc) , specifically so that users never need to take the step of "clear all storage".

It would be good to make sure other specs don't undo the intended benefit of these interventions by assuming "any level of identifier is fine, as long as its reset when the relevant storage bucket is cleared"

**Comment by annevk** [comment]
I don't see how such an intervention is meaningful long term if it doesn't also clear storage.

**Comment by pes10k** [comment]
It does clear storage, it just doesn't clear all storage at the same time.  Safari's ITP, for example, does this, along with Brave, and several extensions.

Its mostly effective for 3p scripts who are riding on the 1p cookie jar

**Comment by annevk** [comment]
I wouldn't call that clearing storage though and it seems circumventable therefore.

**Comment by pes10k** [comment]
It may be circumventable at some margin (as is the case with nearly every privacy protection on the platform), but its also very frequently useful, and so it would be very good to not have standards making the protection more harmful, by introducing unneeded identifiers that will persist across the reduced cookie lifetimes.  That the issue (globally unique device identifiers in the getUserMedia spec) that started the conversation and led to @npdoty filing this issue.

Another way of putting it is just privacy defense in layers; always prefer a more privacy preserving option, all other things being equal.

**Comment by annevk** [comment]
It only works until it's worked around en masse. And meanwhile it makes the situation for web developers rather involved as there is no clear set of principles at work. (Nor is any of it standardized.)

**Comment by pes10k** [comment]
@annevk I don't think we need to argue about whether its an effective protection or not; happy to just note that some browser are using the approach to try and improve privacy.

"Reducing the entropy of device identifiers" as just an application of the general principal of "always prefer the more privacy preserving option, when the privacy preserving is >= useful to users".

Reducing the identifiability of device id's from being globally unique to per eTLD+1 (and dual keyed eTLD+1 on dual keyed platforms) seems like it'd neatly fit that description / principal.

**Comment by hober** [comment]
Given the ongoing work in the Storage Standard, it should (eventually) be clear, at least spec-wise, that "clear all website data" buttons should do what it says on the tin. That said, I do think it would be useful for our Design Principles document to say something like "when adding new storage mechanisms, make sure you hook it up right so that 'clear all website data' does the right thing". Moving this issue to that repo to track that work.
